src/stan/lang/generator/get_verbose_var_type.hpp

```diff
@@ -0,0 +1,36 @@
+#ifndef STAN_LANG_GENERATOR_GET_VERBOSE_VAR_TYPE_HPP
+#define STAN_LANG_GENERATOR_GET_VERBOSE_VAR_TYPE_HPP
+
+#include <stan/lang/ast.hpp>
+#include <string>
+
+namespace stan {
+  namespace lang {
+    /**
+     * Return type string for variable type.
+     *
+     * @param[in] bare_type expression type
+     */
+    std::string
+    get_verbose_var_type(const bare_expr_type bare_type) {
+      bare_expr_type bt(bare_type);
+      if (bt.is_array_type())
+        bt = bt.array_contains();
+
+      if (bt.is_matrix_type()) {
```

Comment:
This is the kind of bad switching that should probably be moved inside the actual `bt` types so that this would look instead like a method on `bt` such as 
```
bt.get_verbose_var_type(const std::string& local_scalar_type);
```
Then the branching on the internal possibilities for bare expression types is encapsulated in the class itself.  The problem here is that when, for example, we add tuples, you'll need to add everything in bare expression type, then find all these branches here to make sure they don't just produce "ill_formed" because the branch of the conditional for the new type never gets implemented.  If the definition is in the local type, it's much more clear what to implement.  Even better if it's pushed down to elements (subclasses or members of a union/variant) so that the compiler will gripe if the condition isn't included.